### DISCUSSION TITLE
Is it possible to use php-mime-mail-parser without composer ?

### DISCUSSION THREAD

**Discussion by Joaquim3** [php-mime-mail-parser/php-mime-mail-parser#376] [Q&A] [answered]
Hello,

First of all Thank you very much for this very nice piece of work. 

I would like to know if I can use the php-mime-mail-parser by directly accessing the classes. I don't want to install composer.

For example, I would to do this:
`<?
require_once "php-mime-mail-parser/src/Parser.php";
$parser = new PhpMimeMailParser\Parser();

$path = 'temp/myemail.eml';
$parser->setPath($path); 
$arrayHeaderTo = $parser->getAddresses('to');
$arrayHeaderFrom = $parser->getAddresses('from');
$subject = $parser->getHeader('subject');
$html = $parser->getMessageBody('html');
echo $html;
?>`

Kind Regards,
Joaquim

**Comment by eXorus** [comment, marked as answer]
Yes of course it's possible like any PHP packages but it's harder to maintain.

**Comment by fgrijze** [comment]
Did you manage it to use php-mime-mail-parser without installing it via composer? If so, can you please tell me how? I'd like to use this class on a virtual host, where I don't have access to SSH/composer.

The mailparse extension is available.

**Reply by Joaquim3** [reply]
Unfortunately, I didn't manage to access the class without composer.

**Reply by eXorus** [reply]
What is the error?

1- Download the last release
2- Put it in a folder
3- Include the class you want to use.

https://stackoverflow.com/questions/40545795/how-do-i-install-composer-php-packages-without-composer

**Reply by eXorus** [reply]
For example with the master version it works:

```php
<?php

require_once 'pmmp/Contracts/ParserInterface.php';
require_once 'pmmp/ParserConfig.php';
require_once 'pmmp/Contracts/CharsetManager.php';
require_once 'pmmp/Charset.php';
require_once 'pmmp/Contracts/ContentTransferEncodingManager.php';
require_once 'pmmp/ContentTransferDecoder.php';
require_once 'pmmp/Contracts/MimeHeaderEncodingManager.php';
require_once 'pmmp/MimeHeaderDecoder.php';
require_once 'pmmp/Contracts/AttachmentInterface.php';
require_once 'pmmp/Attachment.php';
require_once 'pmmp/Contracts/Middleware.php';
require_once 'pmmp/Middleware.php';
require_once 'pmmp/MiddlewareStack.php';
require_once 'pmmp/Entity.php';
require_once 'pmmp/Parser.php';


$email = '
Subject: test email
To: email@example.org
Reply-to: robot@example.org
From: robot@example.org
Message-ID: <1b0ade9e53d92260e2a47762169a17a5@example.org>

test email';


$parser = PhpMimeMailParser\Parser::fromText($email);


echo $parser->getSubject();
```

**Reply by Joaquim3** [reply]
Thanks a lot, eXorus,
That's just what i needed. it works for me too. :-)

**Reply by fgrijze** [reply]
Thx @eXorus! It's working!

**Reply by jasimon9** [reply]
I tried the sample code shown above. I had to make a correction in Entity.php due to an error message. I added #[\ReturnTypeWillChange] above the offsetGet() function. After that, I ran the test code as shown. The attempt to evaluate $parser->getSubject() returned null.

**Reply by jasimon9** [reply]
I tried again and was able to get this test to work, but only with the fromPath() function, not the fromText() function. But looking at the versions of pmmp on Github, I don't see a version that matches the version of PMMP on this page. And I don't know where I got the file set described on this page.

That may not be a problem because it is probably best just to get the latest version from Github anyway. The latest version is 8.0. That version has only 9 files instead of 16 in this version, and the calling sequence is different as well.

So I am wondering what version is being described on this page, and where you can get it by direct download, not Composer.

**Reply by eXorus** [reply]
You are right this code must not be used in production because it uses code from an unreleased version.
Everything is clean now on Github, the main branch or the last release 8.0 are the versions to use.

if you want to use the above code, you have to select the branch next but it is highly not recommended.

To use the last release version without composer, try something like that:

```
<?php

require_once 'pmmp/src/Parser.php';
require_once 'pmmp/Attachment.php';
// ... and all the files included in the src folder.


$email = '
Subject: test email
To: email@example.org
Reply-to: robot@example.org
From: robot@example.org
Message-ID: <1b0ade9e53d92260e2a47762169a17a5@example.org>

test email';

$parser = new Parser();
$parser->setText($email);
echo $parser->getHeader('subject');

```

**Comment by chrisrex** [comment]
Hi,
i try this also without installing composer:
But there is an error:
**Fatal error: Uncaught Error: Call to private PhpMimeMailParser\Parser::__construct() from invalid context**

`<?php
require_once 'Contracts/ParserInterface.php';
require_once 'ParserConfig.php';
require_once 'Contracts/CharsetManager.php';
require_once 'Charset.php';
require_once 'Contracts/ContentTransferEncodingManager.php';
require_once 'ContentTransferDecoder.php';
require_once 'Contracts/MimeHeaderEncodingManager.php';
require_once 'MimeHeaderDecoder.php';
require_once 'Contracts/AttachmentInterface.php';
require_once 'Attachment.php';
require_once 'Contracts/Middleware.php';
require_once 'Middleware.php';
require_once 'MiddlewareStack.php';
require_once 'Entity.php';
require_once 'Parser.php';

$path = 'email.eml';
$parser = new PhpMimeMailParser\Parser();

// 1. Specify a file path (string)
$parser->setPath($path); 

$html = $parser->getMessageBody('html');
// return the html version

echo $html;
?>`

Please help
Best regards
Christoph

**Reply by eXorus** [reply]
As I already said, it works in my example.
I don't know the version you are using here. Be careful the master version is different than the last release.

**Comment by chrisrex** [comment]
Hi,
the subject works but not themessage body:
**Fatal error: Uncaught Error: Call to undefined method PhpMimeMailParser\Parser::getMessageBody()**

require_once 'Contracts/ParserInterface.php';
require_once 'ParserConfig.php';
require_once 'Contracts/CharsetManager.php';
require_once 'Charset.php';
require_once 'Contracts/ContentTransferEncodingManager.php';
require_once 'ContentTransferDecoder.php';
require_once 'Contracts/MimeHeaderEncodingManager.php';
require_once 'MimeHeaderDecoder.php';
require_once 'Contracts/AttachmentInterface.php';
require_once 'Attachment.php';
require_once 'Contracts/Middleware.php';
require_once 'Middleware.php';
require_once 'MiddlewareStack.php';
require_once 'Entity.php';
require_once 'Parser.php';

$path = 'email.eml';
$parser = PhpMimeMailParser\Parser::fromPath($path);

echo $parser->getSubject(); // works fine
echo $parser->getMessageBody('html'); //Error


Best regards Christoph

**Reply by eXorus** [reply]
As I already said, it works in my example.
I don't know the version you are using here. Be careful the master version is different than the last release.

**Reply by rumbleDom** [reply]
Hi chrisrex, have you found a solution?

**Comment by chrisrex** [comment]
Thank you for your answer
Best regards
Christoph

**Comment by eXorus** [comment]
I'm closing this conversation because it's really difficult to understand the answer, because I changed the name of the branch:
- the previous master branch are now the new next branch
- the previous published version are now on main branch
